### PR TITLE
fix error type when handling failures in scheduler

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug
/sig scheduling

#### What this PR does / why we need it:
When we report metrics about scheduling error, we should also update the pod status with `schedulerError`, to keep them consistent.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Pod failed in scheduling due to expected error will be updated with the reason of "SchedulerError" 
rather than "Unschedulable"
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
I don't investigate deeply, but I think this test failure is related to the change, not due to flaky.
Test failed [here](https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/plugins/plugins_test.go#L585-L587) and it waits for the pod condition's reason to become `PodReasonUnschedulable`. Seems related.

https://github.com/kubernetes/kubernetes/blob/6c0bab878c688fe89fee361e811dc30b518a60e2/test/integration/util/util.go#L775

**Comment by kerthcet** [comment]
> PodReasonUnschedulable

Thanks @sanposhiho I'll dig into that.

File: pkg/scheduler/schedule_one.go [hunk 1]
```diff
@@ -150,8 +151,9 @@ func (sched *Scheduler) schedulingCycle(ctx context.Context, state *framework.Cy
 			nominatingInfo = clearNominatedNode
 			klog.ErrorS(err, "Error selecting node for pod", "pod", klog.KObj(pod))
 			metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
+			reason = SchedulerError
 		}
-		sched.FailureHandler(ctx, fwk, podInfo, err, v1.PodReasonUnschedulable, nominatingInfo)
+		sched.FailureHandler(ctx, fwk, podInfo, err, reason, nominatingInfo)
```
Comment (kerthcet) [inline pkg/scheduler/schedule_one.go:156]: This is a break change, since we'll change the reason when updating pod's status. I have no idea whether this was intended, but IMO, I think we should keep the metrics and reason in consistent. cc @Huang-Wei @alculquicondor @ahg-g

Comment (ahg-g) [reply pkg/scheduler/schedule_one.go:156]: Do we consider `reason` values part of the API after https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures?

Does CA rely on this value? I am guessing changing this to a different value is not a an issue with CA since we shouldn't be scaling up because of a scheduling error (and CA will anyway not scale up if it indeed fits)

Comment (Huang-Wei) [reply pkg/scheduler/schedule_one.go:156]: I _guess_ it wouldn't break CA. Its logic to find unschedulabe pods is like: `spec.nodeName==,status.phase!=Succeeded,status.phase!=Failed`:

https://github.com/kubernetes/autoscaler/blob/81d70f94ad2094b2a20f1c8e52d15aff8df7bac1/cluster-autoscaler/utils/kubernetes/listers.go#L183-L194

However, in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-up-work, it says:

> Whenever a Kubernetes scheduler fails to find a place to run a pod, it sets "schedulable" PodCondition to false and reason to "unschedulable".

So cc @x13n for confirmation. (TL;DR, we want to figure out if scheduler updates a pod with an internal ScheduleError reason, would it break CA or not?)

Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:156]: > Do we consider `reason` values part of the API after https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures?

From the description https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#using-pod-statusreason-field, `reason` is not considered as it's arbitrary and hard to track between different versions. cc @alculquicondor @mimowo for confirmation.

Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go:156]: If the reason is in the v1 package, it's considered part of the API. We don't filter based on it for Job failure policy, but that's a specific feature. In any case, we should be good to add new values.

@x13n an error is expected in the cases where apiserver is unavailable (this is already the case), or there are parsing errors #111791 (here we are currently returning Unschedulable). So yes, in general CA should skip these pods. This means that CA is not skipping these errors, which is actually a bug.

File: pkg/scheduler/schedule_one.go [hunk 2]
```diff
@@ -150,8 +151,9 @@ func (sched *Scheduler) schedulingCycle(ctx context.Context, state *framework.Cy
 			nominatingInfo = clearNominatedNode
 			klog.ErrorS(err, "Error selecting node for pod", "pod", klog.KObj(pod))
 			metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
+			reason = SchedulerError
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:154]: we should probably move it to the v1 package.

But I would do it in a separate PR so that we can cherry-pick this as is.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
/hold on @x13n to confirm the current understanding

File: pkg/scheduler/schedule_one.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:154]: Yes, I was plan to.

**Comment by x13n** [comment]
Got it. In that case I think it should be safe to merge this.

/hold cancel
